apps/playercontrol/app.js

```diff
@@ -0,0 +1,116 @@
+      readconfig(function (playerInfo, error) {
+        logger.log('have error ', error)
+        if (playerInfo === null || playerInfo === undefined || playerInfo === '') {
+          speakAndExit(STRING_NO_PLAYER_EXIST)
+          return
+        }
+        logger.log('name = ', playerInfo.name)
+        if (playerInfo.name === null || playerInfo.name === undefined || playerInfo.name === '') {
+          speakAndExit(STRING_NO_PLAYER_EXIST)
+        } else if (playerInfo.name === '@yoda/bluetooth') {
+          activity.voiceCommand('播放蓝牙音乐')
+        } else {
+          logger.log('url = ', playerInfo.url)
+          if (playerInfo.url === null || playerInfo.url === undefined || playerInfo.url === '') {
+            speakAndExit(STRING_NO_PLAYER_EXIST)
+            return
+          }
+          activity.openUrl(playerInfo.url, { preemptive: false })
+        }
+      })
+    }
+  })
+
+  function readconfig (callback) {
+    try {
+      var stat = fs.statSync(DIRPATH)
+      // 为true的话那么存在，如果为false不存在
+      if (stat.isDirectory()) {
+        try {
+          fs.statSync(DATAPATH)
+          // 如果可以执行到这里那么就表示存在了
+          fs.readFile(DATAPATH, 'utf8', function readFileCallback (err, data) {
+            if (err) {
+              logger.error('read data error', err.stack)
+              callback(null, err)
+            }
+            logger.log('data = ', data)
+            callback(JSON.parse(data || '{}'), null)
+          })
+        } catch (e) {
+          // 捕获异常
+          logger.error('read data error1', e)
+          callback(null, e)
+        }
+      }
+    } catch (err) {
+      logger.error('read dir error', err)
+      callback(null, err)
+    }
+  }
+
+  function saveconfig (data) {
+    try {
+      var stat = fs.statSync(DIRPATH)
+      // 为true的话那么存在，如果为false不存在
+      if (stat.isDirectory()) {
+        try {
+          fs.writeFile(DATAPATH, JSON.stringify(data), function (err) {
+            if (err) {
+              logger.error('playercontrol set config: update local data error', err && err.stack)
+            }
+          })
+        } catch (e) {
+          // 捕获异常
+          logger.error('writeFile data error', e)
+        }
+      }
+    } catch (e) {
+      fs.mkdir(DIRPATH, function (error) {
+        if (error) {
+          logger.error(error)
+          return false
+        }
+        logger.log('创建目录成功')
+        fs.writeFile(DATAPATH, JSON.stringify(data), function (err) {
+          if (err) {
+            logger.error('playercontrol set config: update local data error', err && err.stack)
+          }
+        })
+      })
+    }
+  }
+  activity.on('url', urlObj => {
+    logger.log('url is', typeof (urlObj), urlObj)
+    switch (urlObj.pathname) {
+      case '/playercontrol':
+        var saveObj = {}
+        var url = urlObj.href.split('&url=')[1]
+        logger.log('split url is', url)
+        saveObj.name = urlObj.query.name
+        saveObj.url = url
+        saveconfig(saveObj)
```

Comment:
Consider a simple `key-value` pair save mechanism just like Android `shared-preference` to reduce directly manipulation to file system APIs.

Reply:
> values may have long size,so i chose json to save the value

Yes. I mean YodaOS should provide such a mechanism for you.